### PR TITLE
test: events

### LINKED ISSUE #39: Test events
Test events emission

### PR DESCRIPTION
**Pull request by Jean-Grimal**
Fixes #39

### CODE REVIEW

**Review by MathisGD — APPROVED** [review]

**Review by colin-morpho — DISMISSED** [review]

File: test/MorphoTokenEthereumTest.sol [hunk 1]
```diff
@@ -65,6 +69,9 @@ contract MorphoTokenEthereumTest is BaseTest {
 
         deal(address(newMorpho), delegator, amount);
 
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegateeChanged(delegator, address(0), delegatee);
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, 0, amount);
```
Comment (QGarchery) [inline test/MorphoTokenEthereumTest.sol:74]: This pattern does not work: the last line will just be emitted in the test and not be expected to be emitted by a call made later.
```suggestion
        vm.expectEmit(address(newMorpho));
        emit DelegationToken.DelegateeChanged(delegator, address(0), delegatee);
        vm.expectEmit(address(newMorpho));
        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, 0, amount);
```

**Review by QGarchery — CHANGES_REQUESTED** [review]

File: test/MorphoTokenEthereumTest.sol [hunk 2]
```diff
@@ -138,6 +146,9 @@ contract MorphoTokenEthereumTest is BaseTest {
         bytes32 digest = SigUtils.getDelegationTypedDataHash(delegation, address(newMorpho));
         (sig.v, sig.r, sig.s) = vm.sign(privateKey, digest);
 
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegateeChanged(delegator, address(0), delegation.delegatee);
+        emit DelegationToken.DelegatedVotingPowerChanged(delegation.delegatee, 0, amount);
```
Comment (MathisGD) [inline test/MorphoTokenEthereumTest.sol:151]:
```suggestion
        vm.expectEmit(address(newMorpho));
        emit DelegationToken.DelegatedVotingPowerChanged(delegation.delegatee, 0, amount);
```

File: test/MorphoTokenEthereumTest.sol [hunk 3]
```diff
@@ -191,9 +202,15 @@ contract MorphoTokenEthereumTest is BaseTest {
         deal(address(newMorpho), delegator1, amount1);
         deal(address(newMorpho), delegator2, amount2);
 
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegateeChanged(delegator1, address(0), delegatee);
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, 0, amount1);
```
Comment (MathisGD) [inline test/MorphoTokenEthereumTest.sol:207]:
```suggestion
        vm.expectEmit(address(newMorpho));
        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, 0, amount1);
```

File: test/MorphoTokenEthereumTest.sol [hunk 4]
```diff
@@ -191,9 +202,15 @@ contract MorphoTokenEthereumTest is BaseTest {
...
         vm.prank(delegator1);
         newMorpho.delegate(delegatee);
 
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegateeChanged(delegator2, address(0), delegatee);
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, amount1, amount1 + amount2);
```
Comment (MathisGD) [inline test/MorphoTokenEthereumTest.sol:213]:
```suggestion
        vm.expectEmit(address(newMorpho));
        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, amount1, amount1 + amount2);
```

File: test/MorphoTokenEthereumTest.sol [hunk 5]
```diff
@@ -224,6 +241,10 @@ contract MorphoTokenEthereumTest is BaseTest {
 
         vm.startPrank(delegator1);
         newMorpho.delegate(delegatee1);
+
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee1, initialAmount, initialAmount - transferredAmount);
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee2, 0, transferredAmount);
```
Comment (MathisGD) [inline test/MorphoTokenEthereumTest.sol:247]:
```suggestion
        vm.expectEmit(address(newMorpho));
        emit DelegationToken.DelegatedVotingPowerChanged(delegatee2, 0, transferredAmount);
```

File: test/MorphoTokenMigrationTest.sol [hunk 6]
```diff
@@ -93,6 +93,13 @@ contract MorphoTokenEthereumMigrationTest is BaseTest {
 
         vm.startPrank(MORPHO_DAO);
         legacyMorpho.approve(address(bundler), daoTokenAmount);
+
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Transfer(MORPHO_DAO, address(bundler), daoTokenAmount);
+        emit IERC20.Approval(address(bundler), address(wrapper), daoTokenAmount);
+        emit IERC20.Transfer(address(bundler), address(wrapper), daoTokenAmount);
```
Comment (MathisGD) [inline test/MorphoTokenMigrationTest.sol:100]:
```suggestion
        vm.expectEmit(address(newMorpho));
        emit IERC20.Approval(address(bundler), address(wrapper), daoTokenAmount);
        vm.expectEmit(address(newMorpho));
        emit IERC20.Transfer(address(bundler), address(wrapper), daoTokenAmount);
```

File: test/MorphoTokenMigrationTest.sol [hunk 7]
```diff
@@ -115,6 +122,13 @@ contract MorphoTokenEthereumMigrationTest is BaseTest {
 
         vm.startPrank(migrator);
         legacyMorpho.approve(address(bundler), amount);
+
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Transfer(migrator, address(bundler), amount);
+        emit IERC20.Approval(address(bundler), address(wrapper), amount);
+        emit IERC20.Transfer(address(bundler), address(wrapper), amount);
```
Comment (MathisGD) [inline test/MorphoTokenMigrationTest.sol:129]:
```suggestion
        vm.expectEmit(address(newMorpho));
        emit IERC20.Approval(address(bundler), address(wrapper), amount);
        vm.expectEmit(address(newMorpho));
        emit IERC20.Transfer(address(bundler), address(wrapper), amount);
```

File: test/MorphoTokenMigrationTest.sol [hunk 8]
```diff
@@ -93,6 +93,15 @@ contract MorphoTokenEthereumMigrationTest is BaseTest {
 
         vm.startPrank(MORPHO_DAO);
         legacyMorpho.approve(address(bundler), daoTokenAmount);
+
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Transfer(MORPHO_DAO, address(bundler), daoTokenAmount);
+        vm.expectEmit(address(newMorpho));
```
Comment (MathisGD) [inline test/MorphoTokenMigrationTest.sol:99]:
```suggestion
        vm.expectEmit(LEGACY_MORPHO);
```

File: test/MorphoTokenMigrationTest.sol [hunk 9]
```diff
@@ -115,6 +124,15 @@ contract MorphoTokenEthereumMigrationTest is BaseTest {
 
         vm.startPrank(migrator);
         legacyMorpho.approve(address(bundler), amount);
+
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Transfer(migrator, address(bundler), amount);
+        vm.expectEmit(address(newMorpho));
```
Comment (MathisGD) [inline test/MorphoTokenMigrationTest.sol:130]:
```suggestion
        vm.expectEmit(LEGACY_MORPHO);
```

File: test/MorphoTokenMigrationTest.sol [hunk 10]
```diff
@@ -93,6 +93,15 @@ contract MorphoTokenEthereumMigrationTest is BaseTest {
 
         vm.startPrank(MORPHO_DAO);
         legacyMorpho.approve(address(bundler), daoTokenAmount);
+
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Transfer(MORPHO_DAO, address(bundler), daoTokenAmount);
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Approval(address(bundler), address(wrapper), daoTokenAmount);
+        vm.expectEmit(address(newMorpho));
```
Comment (MathisGD) [inline test/MorphoTokenMigrationTest.sol:101]:
```suggestion
        vm.expectEmit(LEGACY_MORPHO);
```

File: test/MorphoTokenMigrationTest.sol [hunk 11]
```diff
@@ -115,6 +124,15 @@ contract MorphoTokenEthereumMigrationTest is BaseTest {
 
         vm.startPrank(migrator);
         legacyMorpho.approve(address(bundler), amount);
+
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Transfer(migrator, address(bundler), amount);
+        vm.expectEmit(LEGACY_MORPHO);
+        emit IERC20.Approval(address(bundler), address(wrapper), amount);
+        vm.expectEmit(address(newMorpho));
```
Comment (MathisGD) [inline test/MorphoTokenMigrationTest.sol:132]:
```suggestion
        vm.expectEmit(LEGACY_MORPHO);
```

**Review by MerlinEgalite — APPROVED** [review]

**Review by MathisGD — APPROVED** [review]

File: test/MorphoTokenEthereumTest.sol [hunk 12]
```diff
@@ -191,9 +203,17 @@ contract MorphoTokenEthereumTest is BaseTest {
...
+        emit DelegationToken.DelegateeChanged(delegator1, address(0), delegatee);
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, 0, amount1);
         vm.prank(delegator1);
         newMorpho.delegate(delegatee);
 
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegateeChanged(delegator2, address(0), delegatee);
+        vm.expectEmit(address(newMorpho));
+        emit DelegationToken.DelegatedVotingPowerChanged(delegatee, amount1, amount1 + amount2);
```
Comment (QGarchery) [inline test/MorphoTokenEthereumTest.sol:216]: Not sure that it's really useful to test events here, because it's already tested in testDelegate

**Review by QGarchery — DISMISSED** [review]

File: test/MorphoTokenEthereumTest.sol (hunk 12 above)
Comment (colin-morpho) [reply test/MorphoTokenEthereumTest.sol:216]: Perhaps, we could group all tests on events in a dedicated test called `testEventsEmited`, the point would be separation of concerns. The logic may be good but doesn't emit the test, so the logic test shouldn't fail if an event is not emited, instead the event test should fail. No strong feeling on this idea though.

**Review by colin-morpho — DISMISSED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Review by MathisGD — APPROVED** [review]

File: test/MorphoTokenOptimismTest.sol [hunk 13]
```diff
@@ -94,6 +99,8 @@ contract MorphoTokenOptimismTest is Test {
         assertEq(morphoOptimism.totalSupply(), 0, "totalSupply");
         assertEq(morphoOptimism.balanceOf(to), 0, "balanceOf(account)");
 
+        vm.expectEmit(address(morphoOptimism));
+        emit DelegationToken.Mint(to, amount);
```
Comment (MathisGD) [inline test/MorphoTokenOptimismTest.sol:103]:
```suggestion
        emit IERC20.Transfer(address(0), to, amount);
```

File: test/MorphoTokenOptimismTest.sol [hunk 14]
```diff
@@ -119,6 +126,9 @@ contract MorphoTokenOptimismTest is Test {
 
         vm.startPrank(BRIDGE);
         morphoOptimism.mint(from, amountMinted);
+
+        vm.expectEmit(address(morphoOptimism));
+        emit DelegationToken.Burn(from, amountBurned);
```
Comment (MathisGD) [inline test/MorphoTokenOptimismTest.sol:131]:
```suggestion
        vm.expectEmit(address(morphoOptimism));
        emit IERC20.Transfer(from, address(0), amount);
```

**Review by MathisGD — APPROVED** [review]

**Pull request merged**